zebra-network/src/address_book.rs

```diff
@@ -65,131 +84,244 @@ impl AddressBook {
     /// Returns the entry corresponding to `addr`, or `None` if it does not exist.
     pub fn get_by_addr(&self, addr: SocketAddr) -> Option<MetaAddr> {
         let _guard = self.span.enter();
-        let (last_seen, services) = self.by_addr.get(&addr).cloned()?;
-        Some(MetaAddr {
-            addr,
-            last_seen,
-            services,
-        })
+        self.by_addr.get(&addr).map(|peer| **peer)
     }
 
-    /// Add `new` to the address book, updating the previous entry if `new` is
-    /// more recent or discarding `new` if it is stale.
-    pub fn update(&mut self, new: MetaAddr) {
+    /// Add `new_peer` to the address book, updating the previous entry if
+    /// `new_peer` is more recent, or discarding `new_peer` if `new_peer` is
+    /// stale.
+    ///
+    /// ## Correctness:
+    ///
+    /// All changes should go through `update` or `take`, to ensure consistency
+    /// and accurate metrics.
```

Comment:
Create an inner type with `update` and `take` methods, and use that type to also do the `Dead` updates.

Reply:
The inner type for #1709 can have an inner enum with an `AliveOrDead` state.

- [ ] map `AliveOrDead`  to `AssumedAlive` or `AssumedDead` as part of the inner-outer API. (Based on the current time.)
- [ ] make the `MetaAddr` state into a template parameter, which defaults to the outer state enum

That way we don't need inner mutability on every query, which is good for correctness and performance.

Also we can:
- [ ] reject updates to the `AssumedDead` and `NeverAttempted` states for existing  peers. (Updates can only be `AttemptPending`, `AssumedAlive`, or `Failed`.)
- [ ] ensure that all inserts of new peers are in the `NeverAttempted` state.